### PR TITLE
i18n: Localize partial routes for logged-out users

### PR DESCRIPTION
**Pull request by claudiucelfilip**
#### Proposed Changes

This allows the `localizeUrl` helper to handle `/partial/url` routes for logged-out users. 
It should replace the [localizeThemesPath](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/themes/helpers.js#L54) and should be used for any future public sections.

#### Testing Instructions
* running `yarn test-packages packages/i18n-utils/src/test/localize-url.js` should not fail
* e2e test should not be affected

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

**Review by yuliyan — COMMENTED** [review]
I don't really like the idea of handling paths within the `localizeUrl` as it currently expects full url as input parameter and is expected to return full url as well. Adding the path names handling, and specifically inside the `INVALID_URL` clause`, could make it a little confusing, imo.

What do you think about creating separate helper to be used just for the path names?

**Review by tyxla — COMMENTED** [review]
I agree with @yuliyan and just wanted to point out that having a seemingly unrelated polymorphic behavior (localizing paths only for logged-out users) can often lead to unexpected bugs, and usually indicate a code smell. Whether we can localize a URL or a path shouldn't be related to whether the user is logged out or not.

**Comment by claudiucelfilip** [comment]
> What do you think about creating separate helper to be used just for the path names?

@yuliyan Sounds good. I moved the functionality to a similar `localizePath` function together with the equivalent custom hook and HOC.

**Comment by tyxla** [comment]
@claudiucelfilip please note that my [note above](https://github.com/Automattic/wp-calypso/pull/66528#pullrequestreview-1071098245) still stands. Have you considered removing the logged in flag and move that to another place (could be another helper function) where the logged out logic is composed on top of the new localize path helper?

**Comment by claudiucelfilip** [comment]
@tyxla I can't argue with [your point](https://github.com/Automattic/wp-calypso/pull/66528#pullrequestreview-1071098245) but this polymorphic is already present in the equivalent [localizeUrl](https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-utils/src/localize-url.tsx#L152). 
We'd need to make the same changes to both localizePath and localizeUrl otherwise it'd become confusing, but I wouldn't inflate the scope of this PR.

**Comment by tyxla** [comment]
Sounds good - I won't be blocking this PR further, but sounds like it would be a good improvement to separate that in another PR, especially now when we have it in more than one place.

File: packages/i18n-utils/src/localize-path.tsx
Comment (lsl) [inline packages/i18n-utils/src/localize-path.tsx]: A subtle design choice exists with `localizeUrl` and `localize_url` where urls that don't need localizing will pass through. It would be good to mirror this design choice. Can we restrict this to `/themes` and `/plugins`?

As an offshoot from that, any url passed in must be a url in the default (english) locale. Any localized paths must be passed through untouched. The `! path.startsWith( `/${ locale }/` );` test here needs to be wider to match any locale. (Although, I think if you're matching `/plugins` and `/themes` directly you won't actually need this check anymore) 

I think this was to make it safe to pass it through multiple times and may not be super relevant on the JS side but I'd prefer to keep these design choices mirrored.

File: packages/i18n-utils/src/test/localize-path.js
Comment (lsl) [inline packages/i18n-utils/src/test/localize-path.js]: e.g. 

```suggestion
		);
		expect( localizePath( '/fr/themes', 'es', true ) ).toEqual( '/fr/themes' );
		expect( localizePath( '/fr/themes', 'es', false ) ).toEqual( '/fr/themes' );
		expect( localizePath( '/some/path', 'es', true ) ).toEqual( '/some/path );
		expect( localizePath( '/some/path', 'es', false ) ).toEqual( '/some/path );
		expect( localizePath( 'https://example.com', 'es', true ) ).toEqual( 'https://example.com' );
		expect( localizePath( 'https://example.com', 'es', false ) ).toEqual( 'https://example.com' );
		expect( localizePath( 'https://wordpress.com/themes', 'es', false ) ).toEqual( 'https://wordpress.com/themes' );
```

**Comment by lsl** [comment]
> I agree with @yuliyan and just wanted to point out that having a seemingly unrelated polymorphic behavior (localizing paths only for logged-out users) can often lead to unexpected bugs, and usually indicate a code smell. Whether we can localize a URL or a path shouldn't be related to whether the user is logged out or not.

Calypso renders different locales under the same url routes and we're reusing those same routes again for logged out use cases. Accounting for logged in state in this code is more or less unavoidable.

**Comment by lsl** [comment]
> I don't really like the idea of handling paths within the localizeUrl as it currently expects full url as input parameter and is expected to return full url as well. What do you think about creating separate helper to be used just for the path names?

I don't like the idea either but I think it might be necessary. The main use case for this function is `href` attributes and given that `href` and `URL()` both support paths as partial urls, extending this function to handle paths seems reasonable to me.

It does add the package based complexity of how to resolve paths in non-calypso contexts. For a far fetched example: If we wanted to use this package on `woocommerce.com` and a `/plugins` route there didn't support localization.

Another issue is that both of these should result in the same location when clicked:

```
localizePath("/themes", "es")
localizeUrl("http://wordpress.com/themes", "es")
```

At the very least localizeUrl should depend on localizePath for the path portion of its execution.

Instead of splitting the exported api into two functions, could we limit scope to paths we know to accept a locale - `/themes` and `/plugins`. Maybe also replacing `INVALID_URL` with `location.origin`, still matching the complete url against e.g. `(wordpress.com|calypso.localhost|wpcal...)/themes` but discarding the origin for path based inputs.

This approach should resolve both issues above but I agree, does make the code quite a bit more complex.

> Adding the path names handling, and specifically inside the INVALID_URL clause, could make it a little confusing, imo.

Pushing "path vs url" confusion up to the caller isn't a great alternative. Could we instead decompose `localizeUrl` into smaller internal functions to manage the complexity better internally?

**Comment by claudiucelfilip** [comment]
@lsl Thanks for the feedback and suggestions. As you suggested, I integrated the relative path handling in the current `localizeUrl` but couldn't get the whole way there. 
> Maybe also replacing INVALID_URL with location.origin, still matching the complete url against

Unfortunately, the `location.origin` is unavailable on the server and using an empty string in the URL object isn't allowed.
Instead, I continued using the INVALID_URL but removed the origin from the output. 
Not sure it's the best approach, but it seems to work well and reuses most of the code for the full URL.

File: packages/i18n-utils/src/localize-url.tsx
Comment (claudiucelfilip) [inline packages/i18n-utils/src/localize-url.tsx]: I haven't been able to get this to work. It seems like assigning an invalid `hostname` is ignored.

Comment (lsl) [reply packages/i18n-utils/src/localize-url.tsx]: Originally came from over here https://github.com/Automattic/wp-calypso/pull/38683/files?w=1 I must have changed it to `hostname = ''` in the move. Could just delete the line?

**Comment by lsl** [comment]
> Unfortunately, the location.origin is unavailable on the server and using an empty string in the URL object isn't allowed.
Instead, I continued using the INVALID_URL but removed the origin from the output.
Not sure it's the best approach, but it seems to work well and reuses most of the code for the full URL.

Sounds like runtime location might be a function dependency. Could we pass location.origin / running context into localizeUrl as a new parameter?

We could do something similar to [provider locale](https://github.com/Automattic/wp-calypso/blob/98fce2af1579334619dd74268f14078c6815cda0/packages/i18n-utils/src/localize-url.tsx#L207) to hide it from calypso callers.

Is there a way we can, when in an SSR context, provide the origin? e.g. adding some kind of origin provider like the locale provider that gets configured in app/boot?

All that sounds pretty gnarly. Is there a better option? 

A calypso specific localizePath implementation is looking like a tempting solution at this point. 

Could we flip this around so we get the single source of truth but still split into a `localizePath` method? e.g. create a `localizePath` function in calypso that looks something like this:
```
function localizePath( path: string, locale?: Locale, isLoggedIn?: boolean ) {
	let url;

	try {
		// Assume all relative paths passed in resolve to a calypso url.
		url = new URL( String( path ), 'https://wordpress.com' );
	} catch ( e ) {
		return path;
	}

	url = localizeUrl( url.toString(), locale, isLoggedIn );

	return url.toString().replace( /^https:\/\/wordpress\.com/, '' );
}
```

This could live in the [i18n utils lib](https://github.com/Automattic/wp-calypso/tree/trunk/client/lib/i18n-utils) instead of the package and already differentiates between node and browser runtimes if you need that still (I don't think you would at this point).

If you want the single localizeUrl function we could also wrap the package one in calypso to add path based inputs handling instead of creating another function call. That would require a find and replace of all useLocalizeUrl imports though.

Thoughts? Other options?

**Comment by claudiucelfilip** [comment]
> We could do something similar to [provider locale](https://github.com/Automattic/wp-calypso/blob/98fce2af1579334619dd74268f14078c6815cda0/packages/i18n-utils/src/localize-url.tsx#L207) to hide it from calypso callers.

@lsl  I had an initial look at how'd this look like andfound that we could make this provider as an opt-in and just pass the origin we want to use. This would avoid adding all the staging environments to the urlLocalizationMapping.
We could also build upon the and also add the logged-in state to not have to pass it every time.
However, all of this was making things harder to follow, more inconsistent, and would require a bit more effort and thought to get everything things running smoothly.

> A calypso specific localizePath implementation is looking like a tempting solution at this point.

I agree. This is probably a better approach for the time being.

> Could we flip this around so we get the single source of truth but still split into a localizePath method

I like the idea; it's fairly simple and straightforward. I've implemented something similar with some extra checks to better match `localizeUrl`.
I tried it on /themes and it worked well out of the box.

File: client/lib/i18n-utils/test/localize-path.js
Comment (lsl) [inline client/lib/i18n-utils/test/localize-path.js]: Neither `localizePath` or `localizeUrl` should be localizing routes by default. Should only be localizing specific routes that support it. (/start, /themes, /plugins, /log-in etc)

```suggestion
		expect( localizePath( '/some/path', 'es', false ) ).toEqual( '/some/path/' );
```

File: client/lib/i18n-utils/localize-path.js
Comment (lsl) [inline client/lib/i18n-utils/localize-path.js]: `localizeUrl` should be covering this situation.

```suggestion
```

File: client/lib/i18n-utils/localize-path.js
Comment (lsl) [inline client/lib/i18n-utils/localize-path.js]: Feels a bit out of scope, if we need for path, do we need this for `localizeUrl` as well?

```suggestion
```

File: client/lib/i18n-utils/localize-path.js
Comment (lsl) [inline client/lib/i18n-utils/localize-path.js]: We should defer to localizeUrl to make these decisions.

```suggestion
	url = localizeUrl( url.toString(), locale, isLoggedIn );
```

**Review by lsl — COMMENTED** [review]
Thanks for the updates, I think this resolves a lot of the previous concerns. 

Left some further suggestions, I don't think we should be doing anything other than wrapping & unwrapping the path in this method. Whether to localize the url or not should be up to `localizeUrl`.

File: client/lib/i18n-utils/test/localize-path.js
Comment (claudiucelfilip) [reply client/lib/i18n-utils/test/localize-path.js]: I'm not sure this should be the intended behaviour. There are a few unspecified URLs  which we still need to be localized:
* [https://wordpress.com/partners/](https://github.com/Automattic/wp-calypso/blob/trunk/client/layout/logged-out.jsx#L81)
* [https://wordpress.com/about/](https://github.com/Automattic/wp-calypso/blob/trunk/client/login/wp-login/index.jsx#L166)
* [https://wordpress.com/webinars](https://github.com/Automattic/wp-calypso/blob/trunk/client/me/help/main.jsx#L201)
* [https://wordpress.com/payments-donations/](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/earn/home.tsx#L274)
* [https://wordpress.com/learn/](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/contextual-help/contextual-help.tsx#L1570)

Although we can always include them to the `urlLocalizationMapping`, moving forward I think this will become hard to maintain. 
For example, there are plans to integrate the LOHP header and footer to Calypso logged-out pages (pau2Xa-4j7-p2). In this case, we'll need to specify and maintain all links to wordpress.com LPs.

File: client/lib/i18n-utils/localize-path.js
Comment (claudiucelfilip) [reply client/lib/i18n-utils/localize-path.js]: Removed

File: client/lib/i18n-utils/localize-path.js
Comment (claudiucelfilip) [reply client/lib/i18n-utils/localize-path.js]: Removed

File: client/lib/i18n-utils/localize-path.js
Comment (claudiucelfilip) [reply client/lib/i18n-utils/localize-path.js]: Removed

**Review by claudiucelfilip — COMMENTED** [review]
I removed all unnecessary path transformations and checks and updated the tests to reflect the changes.

File: client/lib/i18n-utils/test/localize-path.js
Comment (lsl) [reply client/lib/i18n-utils/test/localize-path.js]: > I'm not sure this should be the intended behaviour.

I don't really have a strong opinion about this behavior, mostly just basing it off what `localize_url` and `localizeUrl` was setup to do (be fast and not make assumptions).

If you think you have enough context in calypso to make these assumptions I think you have an argument to add in something separate from `localizeUrl` in `localizePath`.

For LOHP pages it probably makes more sense to add the mappings to localizeUrl and port them to the php `localize_url` function as well.

Comment (claudiucelfilip) [reply client/lib/i18n-utils/test/localize-path.js]: Thinking about it more, the `localizePath` the assumption would be that it receives only calypso related paths. Therefore, it probably should only localize whitelisted paths.

>  think you have an argument to add in something separate from localizeUrl in localizePath.

I found a workaround by setting an empty string for the first item in the `lookup` ([here](https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-utils/src/localize-url.tsx#L194)) array when  `localizeUrl` is called by `localizePath`. This should avoid it matching the `wordpress.com` segment and force it to only localize the whitelisted paths.

File: packages/i18n-utils/src/localize-url.tsx
Comment (lsl) [inline packages/i18n-utils/src/localize-url.tsx]: Could we add tests for this? I'm not sure what skipping the host lookup does.

Comment (claudiucelfilip) [reply packages/i18n-utils/src/localize-url.tsx]: I [pushed](https://github.com/Automattic/wp-calypso/pull/66528/commits/84c485405d4e57ea5f7f81b5b502311c1f77ad53) a few tests. Basically, this parameter prevents host-only path matches such as [wordpress.com](https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-utils/src/localize-url.tsx#L134) because these allow any path to be localized (even for logged-in cases).

File: packages/i18n-utils/src/test/localize-url.js
Comment (lsl) [inline packages/i18n-utils/src/test/localize-url.js]: Thanks for the tests.. this is odd:

```suggestion
		expect( localizeUrl( 'https://wordpress.com/some/path', 'es' ) ).toEqual(
			'https://wordpress.com/some/path/'
		);
```

This test shouldn't be failing but it is even without your change, I thought this function was safe to run on variables containing unknown urls but that doesn't look to be the case for WordPress.com urls.

The localizeUrl function seems to exhibit the behavior you need for WordPress.com urls already. Maybe you don't need skipHostLookup after all?

Comment (lsl) [reply packages/i18n-utils/src/test/localize-url.js]: @yuliyan is this a bug in `localizeUrl` or do all WordPress.com paths support localizing now? Feel free to correct my assumptions here, my memory could be way off base. :shrug:

Comment (claudiucelfilip) [reply packages/i18n-utils/src/test/localize-url.js]: > The localizeUrl function seems to exhibit the behavior you need for WordPress.com urls already. Maybe you don't need skipHostLookup after all?

I don't think we want the same behavior for localizePath. 
For example, without `skipHostLookup` running localizePath('/some/path') would localize it even though the path is not whitelisted and even if the user is logged in.

Comment (lsl) [reply packages/i18n-utils/src/test/localize-url.js]: > I don't think we want the same behavior for localizePath.

Definitely want the same behavior between the two (not necessarily whats there now). I should be able to take any wordpress.com url in calypso and wrap it in localizeUrl without having to check if it's working right.

Comment (yuliyan) [reply packages/i18n-utils/src/test/localize-url.js]: > @yuliyan is this a bug in localizeUrl or do all WordPress.com paths support localizing now? Feel free to correct my assumptions here, my memory could be way off base. 🤷

@lsl, IIRC, we localize all paths on WordPress.com since landing pages have been moved to sub directories.

The equivalent function for localizing URLs, that we have on the back-end, would also prefix `https://wordpress.com/some/path` with the locale slug, i.e. `https://wordpress.com/es/some/path`.

Comment (lsl) [reply packages/i18n-utils/src/test/localize-url.js]: Would that mean using localizeUrl inside localizePath is a mistake? It sounds like localizeUrl doesn't cover calypso routes?

Comment (yuliyan) [reply packages/i18n-utils/src/test/localize-url.js]: > It sounds like localizeUrl doesn't cover calypso routes?

Currently, it has URL localization mapping for `/theme`, `/themes`, `log-in`. After a quick audit, I found it lacks of mapping for `/start` and `/accept-invite` (?) routes. These should be all current Calypso routes that accept locale parameter.

As for whether it's a mistake or not - to me it seems a bit odd that `localizePath` would localize non-Calypso paths (e.g. `/support`, `/forums`, etc.), as it's currently one of the possible code paths. However, I can't really tell if that's an actual concern, given that `localizePath` should never be used for non-Calypso paths in the first place.

**Review by yuliyan — COMMENTED** [review]
I did some testing and came across the following problem - both `localizePath( '/wp-admin', 'es', false )` and `localizePath( '/wp-login.php', 'es', false )` return a full URL string as a result (`https://es.wordpress.com/wp-admin/` and `https://es.wordpress.com/wp-login.php` respectively), rather than the expected path name string.

Since these are invalid Calypso paths, I'd expect to have the original input path returned, rather than having any localization or other transformations being applied. However, the current approach makes it that all `wordpress.com/*` URL mappings could be a possible match which leads to such unexpected behaviour.

I think we should be more explicit on which `localizeUrl` mapping rules should be filtered when used within `localizePath`. In order to do that, we should define [the URLs that are associated with valid Calypso paths](https://github.com/Automattic/wp-calypso/blob/84c485405d4e57ea5f7f81b5b502311c1f77ad53/packages/i18n-utils/src/localize-url.tsx#L145-L153).

File: packages/i18n-utils/src/test/localize-url.js
Comment (lsl) [reply packages/i18n-utils/src/test/localize-url.js]: > These should be all current Calypso routes that accept locale parameter.

Thanks!

> As for whether it's a mistake or not - to me it seems a bit odd that localizePath would localize non-Calypso paths (e.g. /support, /forums, etc.), as it's currently one of the possible code paths. However, I can't really tell if that's an actual concern, given that localizePath should never be used for non-Calypso paths in the first place.

Yea this is pretty hairy. Not sure what the best approach is.

**Comment by lsl** [comment]
> I think we should be more explicit on which localizeUrl mapping rules should be filtered when used within localizePath. In order to do that, we should define [the URLs that are associated with valid Calypso paths](https://github.com/Automattic/wp-calypso/blob/84c485405d4e57ea5f7f81b5b502311c1f77ad53/packages/i18n-utils/src/localize-url.tsx#L145-L153).

Just to make sure I understand this right, instead of adding `skipHostLookup`, specify the routes needed in localizeUrl?

**Comment by yuliyan** [comment]
> > I think we should be more explicit on which localizeUrl mapping rules should be filtered when used within localizePath. In order to do that, we should define [the URLs that are associated with valid Calypso paths](https://github.com/Automattic/wp-calypso/blob/84c485405d4e57ea5f7f81b5b502311c1f77ad53/packages/i18n-utils/src/localize-url.tsx#L145-L153).
> 
> Just to make sure I understand this right, instead of adding `skipHostLookup`, specify the routes needed in localizeUrl?

Yes, either specify the routes individually, or group the URLs (e.g. `calypso`, `wpcom`, `jetpack`) and specify the group. Do we have other options?

**Comment by claudiucelfilip** [comment]
> Yes, either specify the routes individually, or group the URLs (e.g. `calypso`, `wpcom`, `jetpack`) and specify the group. Do we have other options?

What do you think of using the [sections.js](https://github.com/Automattic/wp-calypso/blob/trunk/client/sections.js#L220) and checking for path matches for sections with `enableLoggedOut`?

**Comment by yuliyan** [comment]
> > Yes, either specify the routes individually, or group the URLs (e.g. `calypso`, `wpcom`, `jetpack`) and specify the group. Do we have other options?
> 
> What do you think of using the [sections.js](https://github.com/Automattic/wp-calypso/blob/trunk/client/sections.js#L220) and checking for path matches for sections with `enableLoggedOut`?

I thought of that, but my concern is that not all sections are defined in that module - e.g. [/log-in](https://github.com/Automattic/wp-calypso/blob/trunk/client/login/index.web.js#L23), [/new](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/gutenboarding/section.ts), and [/setup](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/section.ts).

**Comment by claudiucelfilip** [comment]
So far it seems that wrapping around `localizeUrl` is more complicated than it's worth.

Playing around I found (and pushed) an approach where we use a dedicated path list for `localizePath` but can also reuse any items from the `urlLocalizationMapping` list. 
I think this offers enough flexibility to define specific calypso paths but not have to duplicate functionality.

File: client/lib/i18n-utils/localize-path.js
Comment (lsl) [inline client/lib/i18n-utils/localize-path.js]: Should this include `/start/`? (Don't know, just know there is /start and /new and both should be localized / logged-out-able)

File: client/lib/i18n-utils/localize-path.js
Comment (lsl) [inline client/lib/i18n-utils/localize-path.js:32]: Did you mean to add plugins routes in here too?

File: client/lib/i18n-utils/test/localize-path.js
Comment (lsl) [inline client/lib/i18n-utils/test/localize-path.js]:
```suggestion
		expect( localizePath( '/start/user', 'es', false ) ).toEqual( '/start/user/es' );
		expect( localizePath( '/start/user/', 'es', false ) ).toEqual( '/start/user/es/' );
```
`localizePath` shouldn't be adding trailing slashes (we redirect to strip them), but lets not remove them if provided either.

**Review by lsl — APPROVED** [review]
LGTM, commented a couple of minor changes you might want to make but nothing blocking.

File: client/lib/i18n-utils/test/localize-path.js
Comment (claudiucelfilip) [reply client/lib/i18n-utils/test/localize-path.js]: Sounds good! 
I pushed a fix for this and updated the rest of the tests.

File: client/lib/i18n-utils/localize-path.js
Comment (claudiucelfilip) [reply client/lib/i18n-utils/localize-path.js:32]: Good idea, now that #66169 has been deployed.
I'll follow up with a PR to use this inside Plugins and Themes.

**Pull request merged**
